### PR TITLE
406 feature add sqlx support to interact with the postgres db

### LINKED ISSUE #406: [Feature]: Add sqlx support to interact with the Postgres DB
### Describe the new feature

We aim to integrate sqlx into our backend to connect with the Postgres DB. sqlx is a SQL toolkit that provides compile-time checked SQL queries without sacrificing the benefits of Rust’s type safety.

### Describe the solution you'd like

- Add sqlx to the dependencies in Cargo.toml
- Ensure we use the appropriate sqlx postgres features
- Configure connection settings, such as maximum connections and timeout settings.
- Check the connection with the Postgres DB host
- Write unit and integration tests to verify the correct implementation of sqlx features.

### Describe alternatives you've considered

- **Diesel**: A popular ORM for Rust that offers strong type safety and compile-time checks. However, sqlx was chosen for its flexibility in handling raw SQL and asynchronous support without requiring macros.
- **SeaORM**: A newer ORM in the Rust ecosystem that also offers type safety and async support. sqlx was preferred due to its maturity and larger community support.

### Additional context/resources
[sqlx repository](https://github.com/launchbadge/sqlx)
[Realworld example of Axum with sqlx](https://github.com/davidpdrsn/realworld-axum-sqlx)

### PR DESCRIPTION
**Pull request by mattiapitossi**
Closes #406

### CODE REVIEW

File: dcapal-backend/docker-compose.yml [hunk 1]
```diff
@@ -17,6 +17,44 @@ services:
...
+      PGPORT: 5432
+      POSTGRES_PORT: 5432
+      PGPASSWORD: postgres
+      POSTGRES_PASSWORD: postgres
+      PGDATABASE: postgres
+      POSTGRES_DB: postgres
+    volumes:
+      - ./volumes/db/realtime.sql:/docker-entrypoint-initdb.d/migrations/99-realtime.sql:Z
+      # Must be superuser to create event trigger
+      - ./volumes/db/webhooks.sql:/docker-entrypoint-initdb.d/init-scripts/98-webhooks.sql:Z
```
Comment (leonardoarcari) [inline dcapal-backend/docker-compose.yml:44]: This mounted file seems to fail when initializing the container. Check out the `e2e` CI job

File: dcapal-backend/src/lib.rs [hunk 2]
```diff
@@ -100,6 +104,10 @@ impl DcaServer {
 
         let redis = build_redis_pool(&config.server.redis)?;
 
+        let postgres = build_postgres_pool(&config.server.postgres).await?;
+
+        sqlx::migrate!().run(&postgres).await?;
```
Comment (leonardoarcari) [inline dcapal-backend/src/lib.rs:109]: Think we should migrate the DB outside of DcaPal server app to have better control over it. Something like a job that runs after Postgres container is healthy. Then we make DcaPal container depend on it so we don't start until DB has been migrated.

Something like this but waiting for Postgres healthiness before we run it: https://stackoverflow.com/a/64153124

**Review by leonardoarcari — CHANGES_REQUESTED** [review]

File: dcapal-backend/src/lib.rs (hunk 2 above)
Comment (mattiapitossi) [reply dcapal-backend/src/lib.rs:110]: It's a good point. At the moment, the Postgres docker instance is for test and local development purposes only, as we are using the AWS Supabase instance which should be always up and running. I've added a TODO for when we switch to the docker container to run the migration from the script. Let me know if it makes sense for you

File: dcapal-backend/docker-compose.yml (hunk 1 above)
Comment (mattiapitossi) [reply dcapal-backend/docker-compose.yml]: The pipeline is now fixed. I've replaced the Supabase docker configuration with plain Postgres so we don't have to rely upon them in the future if we decide to migrate.

**Review by leonardoarcari — CHANGES_REQUESTED** [review]

File: dcapal-backend/src/lib.rs (hunk 2 above)
Comment (leonardoarcari) [reply dcapal-backend/src/lib.rs:110]: Yeah I agree for the healthcheck part 👍 But for the migration part, imo it's an anti-pattern to run it as part of the main application.

When you make any DB change, I find it less error prone to run the `sqlx migrate` step against the AWS instance, verify it's all good and then deploy the new image.

Given how small time we have to maintain the app, I don't want to risk unintentional migrations happening across restarts.

Do you mind adding a migrate script similar to the `start.sh` you provided? `sqlx` should be smart enough to pick up any `.env` file in the same folder and use the Postgres connection params to connect to the AWS/local instance.

File: dcapal-backend/docker-compose.yml (hunk 1 above)
Comment (leonardoarcari) [reply dcapal-backend/docker-compose.yml]: Awesome! 🙌

**Comment by mattiapitossi** [comment]
Closing as these features will be added in #431

**Pull request closed without merging**